specs/public_inputs.md

```diff
@@ -0,0 +1,297 @@
+
+The following diagram shows the the public input approach using the RLC shortcut:
+
+![](./public_inputs.rev1.png)
+
+## Verify a KZG BLS commitment inside a circuit
+
+Here's a proposal to prove that the commitment using KZG with a BLS curve (as
+defined in EIP 4844) uses the same values as the ones found in an advice column
+of a circuit:
+
+1. Pick random `x`
+2. Evaluate the polynomial used in `commitment_bls` at `x` and get `y`.  So
+   `verify_kzg_bls_proof(commitment_bls, x, y, quotient_kzg) == True`, where
+   `y` is in modulus of BLS
+2. Pass `(x, y)` into the aggregation circuit (as public inputs)
+3. Pass `(x, y)` into the `public_inputs_circuit` (as public inputs)
+4. `public_inputs_circuit` contains a column with `raw_public_inputs`
+5. Inside the circuit, evaluate the polynomial defined with `raw_public_inputs`
+   as its Lagrange coefficients in the BLS modulus at `x` and verify that the
+   result is `y`.
+   We use the [barycentric formula](https://en.wikipedia.org/wiki/Lagrange_polynomial#Barycentric_form)
+   to evaluate the polynomial using its Lagrange coefficients efficiently.
+
+How to pick the random challenge `x`:
+- `x = hash(commitment_bls(raw_public_inputs) || poly_commitment(PublicInputsCircuit:raw_public_inputs))`
+
+This is because:
+- The prover shouldn't know x before the `commitment_bls` is calculated
+- The prover shouldn't know x before the witness `PublicInputsCircuit:raw_public_inputs` is committed
+- In summary: the prover shouldn't be able to change any of the commitments
+  after it learns about `x`, otherwise the prover is able to construct a polynomial
+  with values different than `raw_public_inputs` that evaluates to `y` on `x`.
+
+# PublicInputs Circuit
+
+## Setup
+
+All the necessary public data is arranged in a single array of elements (called
+`raw_public_inputs`), following the layout of the `block_table` `value` column,
+`tx_table` {`tx_id`, `index`, `value`} columns, and extra fields in between.
+
+## Public Inputs
+
+- `rand_rpi`: Randomness used to "compress" the raw public inputs
+- `rpi_rlc`: Random Linear Combination of the raw public inputs (using `rand_rpi` as randomness)
+- `chain_id`: Chain ID, used to match the Chain ID public input used in the Tx Circuit
+- `state_root`: State Root of current block, used to match the State Root of current block public input used in the State Circuit
+- `state_root_prev`: State Root of previous block, used to match the State Root of previous block public input used in the State Circuit
+
+## Behaviour
+
+First, the circuit calculates the Random Linear Combination of a column
+containing the raw public inputs array, and verifies that the result matches
+the `rpi_rlc` passed via public inputs using `rand_rpi` as randomness.
+
+Second, the circuit proves that the contained `block_table -> value` and
+`tx_table -> {tx_id, index, value}` columns correspond to the correct sections
+of the raw public inputs column.
```

Comment:
I thought about 2 ways to do that.  One is with copy constraints, and the other is with rotations and constraints of the type `block_table_value[i] - raw_public_input[i - offset] = 0`.
I haven't explained which approach to use in the markdown document because I think that's an implementation detail!

In the python implementation, I've done the second option (as suggested by Han for being the cheapest option).  Nevertheless, once we implement this it shouldn't be very hard to change this if the other option gives better results in the benchmarks.

Reply:
> I checked the impl. But I thought that maybe since we;re indeed checking subsets, we could do a shuffling proof.

I'm not very familiar with the shuffling argument and the API to use it.  Also, in this case, the order of items is equal, so there's no actual shuffle.

This is a diagram of what's intended to happen: 
![image](https://user-images.githubusercontent.com/24987329/173331265-77384fa2-db12-4d86-a279-8043baa146f8.png)


> Anyways, the copy constraints look correct to me. Just wondering what will happen if the table grows/shrinks. Will the copy constraints disappear?

Notice that the python implementation is not using copy constraints but rotations.

All the table sizes are static based of constant parameters, and the offsets used in copy constraints positions / rotation offsets are based on those constants; so there should be no problem :)

Reply:
Oh right!! I confused with the copies of the ChainID and the other params!

Looks fine then! The shuffle argument is basically what is used in the lookup argument in halo2 plonk for example where we check that a collection of data is a subset of another one. 
I think @han0110 made some implementation inside of the circuit on his PR https://github.com/zcash/halo2/pull/593

Specifically this: https://github.com/zcash/halo2/blob/ecd70819c571d44a960f29969c9d751811d91263/halo2_proofs/examples/shuffle.rs